pkg/gardenlet/operation/botanist/secrets.go

```diff
@@ -495,29 +492,48 @@ func (b *Botanist) DeployCloudProviderSecret(ctx context.Context) error {
 		if err != nil {
 			return err
 		}
-		additionalAnnotations = map[string]string{
-			securityv1alpha1constants.AnnotationWorkloadIdentityNamespace:     credentials.Namespace,
-			securityv1alpha1constants.AnnotationWorkloadIdentityName:          credentials.Name,
-			securityv1alpha1constants.AnnotationWorkloadIdentityContextObject: string(shootData),
-		}
-		additionalLabels = map[string]string{
-			securityv1alpha1constants.LabelPurpose:                  securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor,
-			securityv1alpha1constants.LabelWorkloadIdentityProvider: credentials.Spec.TargetSystem.Type,
+
+		mutateFunc = func() error {
+			secret.Annotations = utils.MergeStringMaps(secret.Annotations,
+				map[string]string{
+					securityv1alpha1constants.AnnotationWorkloadIdentityNamespace:     credentials.Namespace,
+					securityv1alpha1constants.AnnotationWorkloadIdentityName:          credentials.Name,
+					securityv1alpha1constants.AnnotationWorkloadIdentityContextObject: string(shootData),
+				},
+			)
+			secret.Labels = utils.MergeStringMaps(
+				secret.Labels,
+				map[string]string{
+					securityv1alpha1constants.LabelPurpose:                  securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor,
+					securityv1alpha1constants.LabelWorkloadIdentityProvider: credentials.Spec.TargetSystem.Type,
+				},
+				map[string]string{v1beta1constants.GardenerPurpose: v1beta1constants.SecretNameCloudProvider},
+			)
+
+			// Preserve the token field if already set, otherwise it is removed
+			// which triggers unwanted renewal of the workload identity token.
+			if token, ok := secret.Data["token"]; ok {
+				data["token"] = token
+			}
+
+			secret.Type = corev1.SecretTypeOpaque
+			secret.Data = utils.MergeStringMaps(secret.Data, data)
```

Comment:
Why do we merge the old and new fields? Can't we just do the following?

```suggestion
			secret.Data = data
```

Reply:
Good catch, this is left over from the first version of fix.